src/jit/morph.cpp

```diff
@@ -7237,17 +7237,511 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+// identified as a potential tailcall to an actual tailcall.
+//
+// Arguments:
+//    call - The call to morph.
+//    newNode - The new node to use if this function returns true.
+//
+// Return Value:
+//    Returns true or false based on whether the call was morphed into a tailcall.
+//    If this function returns true the call is done being morphed and the new node
+//    should be used. Otherwise the call will be demoted to a regular call and
+//    should go through normal morph.
+//
+// Notes:
+//    This is called only for calls that the importer has already identified as
+//    potential tailcalls. It will do profitability and legality checks and
+//    classify which kind of tailcall we are able to (or should) do, along with
+//    modifying the trees to perform that kind of tailcall.
+//
+bool Compiler::fgMorphPotentialTailCall(GenTreeCall* call, GenTree** newNode)
+{
+    // It should either be an explicit (i.e. tail prefixed) or an implicit tail call
+    assert(call->IsTailPrefixedCall() ^ call->IsImplicitTailCall());
+
+    // It cannot be an inline candidate
+    assert(!call->IsInlineCandidate());
+
+    auto failTailCall = [&](const char* reason) {
+#ifdef DEBUG
+        if (verbose)
+        {
+            printf("\nRejecting tail call in morph for call ");
+            printTreeID(call);
+            printf(": %s\n", reason);
+        }
+#endif
+
+        // for non user funcs, we have no handles to report
+        info.compCompHnd->reportTailCallDecision(nullptr,
+                                                 (call->gtCallType == CT_USER_FUNC) ? call->gtCallMethHnd : nullptr,
+                                                 call->IsTailPrefixedCall(), TAILCALL_FAIL, reason);
+
+        // We have checked the candidate so demote.
+        call->gtCallMoreFlags &= ~GTF_CALL_M_EXPLICIT_TAILCALL;
+#if FEATURE_TAILCALL_OPT
+        call->gtCallMoreFlags &= ~GTF_CALL_M_IMPLICIT_TAILCALL;
+#endif
+    };
+
+    if (call->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC)
+    {
+        failTailCall("Might turn into an intrinsic");
+        return false;
+    }
+
+    if (opts.compNeedSecurityCheck)
+    {
+        failTailCall("Needs security check");
+        return false;
+    }
+
+    if (compLocallocUsed || compLocallocOptimized)
+    {
+        failTailCall("Localloc used");
+        return false;
+    }
+
+#ifdef _TARGET_AMD64_
+    // Needed for Jit64 compat.
+    // In future, enabling tail calls from methods that need GS cookie check
+    // would require codegen side work to emit GS cookie check before a tail
+    // call.
+    if (getNeedsGSSecurityCookie())
+    {
+        failTailCall("GS Security cookie check");
+        return false;
+    }
+#endif
+
+#ifdef DEBUG
+    // DDB 99324: Just disable tailcall under compGcChecks stress mode.
+    if (opts.compGcChecks)
+    {
+        failTailCall("GcChecks");
+        return false;
+    }
+#endif
+
+    // We have to ensure to pass the incoming retValBuf as the
```

Comment:
I realize that this is pre-existing code, but I don't know under what conditions we'd have a `retValBuf` node that wasn't `compRetBuffArg`. If you could explain that situation here that would be an improvement.

Reply:
Looks like this can happen due to inlining. It happens in [this switch](https://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/Advanced/ColorTranslator.cs,143) in `ColorTranslator.FromOle` when `FromKnownColor` is inlined. First `FromKnownColor` is marked as a tailcall candidate, and then also the inlined `FromName` is marked as a tailcall candidate, and we end up with this tree:

```
***** BB38, stmt 15
               [000388] ------------              *  STMT      void  (IL 0x099...  ???)
               [000381] S-C-G-------              \--*  CALL      void   System.Drawing.Color.FromName
               [000386] ------------ arg0            +--*  ADDR      byref 
               [000385] ------------                 |  \--*  LCL_VAR   struct V03 tmp1         
               [000364] --C-G------- arg1            \--*  CALL      ref    System.Enum.ToString
               [000378] ------------ this in rcx        \--*  BOX       ref   
               [000377] ------------                       \--*  LCL_VAR   ref    V05 tmp3         

------------ BB39 [???..???) (return), preds={} succs={}

***** BB39, stmt 16
               [000243] ------------              *  STMT      void  (IL   ???...  ???)
               [000392] -A----------              \--*  ASG       struct (copy)
               [000391] ------------                 +--*  OBJ(24)   struct
               [000389] ------------                 |  \--*  LCL_VAR   byref  V00 RetBuf       
               [000390] ------------                 \--*  LCL_VAR   struct V03 tmp1         

***** BB39, stmt 17
               [000246] ------------              *  STMT      void  (IL   ???...  ???)
               [000245] ------------              \--*  RETURN    byref 
               [000244] ------------                 \--*  LCL_VAR   byref  V00 RetBuf   
```

I will add a comment.

Reply:
Thanks - a comment would be great!